### PR TITLE
Change BartLearnedPositionalEmbedding's forward method signature to support Opacus training

### LINKED ISSUE #18425: BartLearnedPositionalEmbedding's forward method signature obstructs private (Opacus) training of BART
### System Info

-`transformers` version: 4.20.1
-Platform: Linux-5.4.0-1086-azure-x86_64-with-glibc2.17
-Python version: 3.8.13
-Huggingface_hub version: 0.8.1
-PyTorch version (GPU?): 1.9.1+cu102 (False)
-Tensorflow version (GPU?): not installed (NA)
-Flax version (CPU?/GPU?/TPU?): not installed (NA)
-Jax version: not installed
-JaxLib version: not installed
-Using GPU in script?: yes (NA)
-Using distributed or parallel set-up in script?: no (NA)

### Who can help?

Tagging @patil-suraj as BART model owner.

Details:
The signature of `BartLearnedPositionalEmbedding`'s forward method takes an input of type `torch.Size`, which breaks in Opacus. The reason is that Opacus makes a (reasonable) assumption that all layers take input of type `torch.Tensor`. 

In particular, opacus/grad_sample/grad_sample_module.py line 190 (the `capture_activations_hook` method) tries to detach the input from device via:

`module.activations.append(forward_input[0].detach())`

If we pass the tensor instead, this will allow fine-tuning BART-type summarization models with differential privacy.

Only a few lines of code need to be changed in `modeling_bart.py`. In particular, the forward signature of `BartLearnedPositionalEmbedding.forward()` and references to this method.

I already have a change implemented with BART-related tests passing. More than happy to create a PR which I can tag you in @patil-suraj.

### Information

- [X] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

```
import torch

from transformers.models.bart.modeling_bart import  BartLearnedPositionalEmbedding


from opacus.tests.grad_samples.common import GradSampleHooks_test
class TestPositionalEmbedding(GradSampleHooks_test):
    def test_grad_sample(self):
        """
        Verify that our custom implementation of the grad sample for huggingface's Conv1D
        layer works. We largely build on the test routines in opacus's library.
        """
        register_grad_sampler()
        batch_size = 1
        max_pos_embs = 10
        embed_dim = 3
    
        # x = torch.randint(0, max_pos_embs - 1, (batch_size, embed_dim))
        x = torch.tensor([0,1,2,3,4,5,6,7,8,9]).reshape(batch_size, max_pos_embs)
        layer = BartLearnedPositionalEmbedding(max_pos_embs, embed_dim)
        self.run_test(x, layer, batch_first=True)
```
where a custom `register_grad_sampler()` method is called for `BartLearnedPositionalEmbedding` layer.

### Expected behavior

Test above should pass.

### PR DESCRIPTION
**Pull request by donebydan**
As outlined in #18425, this PR changes the signature of `BartLearnedPositionalEmbedding`'s forward method signature to take the `input_ids` tensor (and not just its shape). This is needed to enable private training of BART via DP-SGD in Opacus. PR welcomed by @sgugger in linked issue.

Fixes #18425.

### CODE REVIEW

**Comment by sgugger** [comment]
You will also need to apply the same changes to all the models that are touched by the change in embedding (mBART, plBARt etc) to have the tests passing.

**Comment by donebydan** [comment]
@sgugger I am iterating through. 👍 thanks for the heads up, though!

**Comment by sgugger** [comment]
Ah, this also looks like it's breaking the conversion to `torch.fx`. Let's see if @michaelbenayoun can think of an easy solution to that.

**Comment by donebydan** [comment]
Thanks, I was about to ask.. any thoughts? I am not well-versed in torch's symbolic tracer (or FX generally). I'm happy to do the work if you can point me somewhere useful 😄

**Comment by donebydan** [comment]
Current offending line is line 985 in `src/transformers/utils/fx.py` (`HFTracer.trace()`):
`self.graph = super().trace(root, concrete_args=concrete_args)` 

where `root` is 
```
PLBartModel(
  (shared): Embedding(99, 16, padding_idx=1)
  (encoder): PLBartEncoder(
    (embed_tokens): Embedding(99, 16, padding_idx=1)
    (embed_positions): PLBartLearnedPositionalEmbedding(102, 16)
    (layers): ModuleList(
      (0): PLBartEncoderLayer(
        (self_attn): PLBartAttention(
          (k_proj): Linear(in_features=16, out_features=16, bias=True)
          (v_proj): Linear(in_features=16, out_features=16, bias=True)
          (q_proj): Linear(in_features=16, out_features=16, bias=True)
          (out_proj): Linear(in_features=16, out_features=16, bias=True)
        )
        (self_attn_layer_norm): LayerNorm((16,), eps=1e-05, elementwise_affine=True)
        (activation_fn): GELUActivation()
        (fc1): Linear(in_features=16, out_features=4, bias=True)
        (fc2): Linear(in_features=4, out_features=16, bias=True)
        (final_layer_norm): LayerNorm((16,), eps=1e-05, elementwise_affine=True)
      )
      (1): PLBartEncoderLayer(
        (self_attn): PLBartAttention(
          (k_proj): Linear(in_features=16, out_features=16, bias=True)
          (v_proj): Linear(in_features=16, out_features=16, bias=True)
          (q_proj): Linear(in_features=16, out_features=16, bias=True)
          (out_proj): Linear(in_features=16, out_features=16, bias=True)
        )
        (self_attn_layer_norm): LayerNorm((16,), eps=1e-05, elementwise_affine=True)
        (activation_fn): GELUActivation()
        (fc1): Linear(in_features=16, out_features=4, bias=True)
        (fc2): Linear(in_features=4, out_features=16, bias=True)
        (final_layer_norm): LayerNorm((16,), eps=1e-05, elementwise_affine=True)
      )
    )
    (layernorm_embedding): LayerNorm((16,), eps=1e-05, elementwise_affine=True)
  )
  (decoder): PLBartDecoder(
    (embed_tokens): Embedding(99, 16, padding_idx=1)
    (embed_positions): PLBartLearnedPositionalEmbedding(102, 16)
    (layers): ModuleList(
      (0): PLBartDecoderLayer(
        (self_attn): PLBartAttention(
          (k_proj): Linear(in_features=16, out_features=16, bias=True)
          (v_proj): Linear(in_features=16, out_features=16, bias=True)
          (q_proj): Linear(in_features=16, out_features=16, bias=True)
          (out_proj): Linear(in_features=16, out_features=16, bias=True)
        )
        (activation_fn): GELUActivation()
        (self_attn_layer_norm): LayerNorm((16,), eps=1e-05, elementwise_affine=True)
        (encoder_attn): PLBartAttention(
          (k_proj): Linear(in_features=16, out_features=16, bias=True)
          (v_proj): Linear(in_features=16, out_features=16, bias=True)
          (q_proj): Linear(in_features=16, out_features=16, bias=True)
          (out_proj): Linear(in_features=16, out_features=16, bias=True)
        )
        (encoder_attn_layer_norm): LayerNorm((16,), eps=1e-05, elementwise_affine=True)
        (fc1): Linear(in_features=16, out_features=4, bias=True)
        (fc2): Linear(in_features=4, out_features=16, bias=True)
        (final_layer_norm): LayerNorm((16,), eps=1e-05, elementwise_affine=True)
      )
      (1): PLBartDecoderLayer(
        (self_attn): PLBartAttention(
          (k_proj): Linear(in_features=16, out_features=16, bias=True)
          (v_proj): Linear(in_features=16, out_features=16, bias=True)
          (q_proj): Linear(in_features=16, out_features=16, bias=True)
          (out_proj): Linear(in_features=16, out_features=16, bias=True)
        )
        (activation_fn): GELUActivation()
        (self_attn_layer_norm): LayerNorm((16,), eps=1e-05, elementwise_affine=True)
        (encoder_attn): PLBartAttention(
          (k_proj): Linear(in_features=16, out_features=16, bias=True)
          (v_proj): Linear(in_features=16, out_features=16, bias=True)
          (q_proj): Linear(in_features=16, out_features=16, bias=True)
          (out_proj): Linear(in_features=16, out_features=16, bias=True)
        )
        (encoder_attn_layer_norm): LayerNorm((16,), eps=1e-05, elementwise_affine=True)
        (fc1): Linear(in_features=16, out_features=4, bias=True)
        (fc2): Linear(in_features=4, out_features=16, bias=True)
        (final_layer_norm): LayerNorm((16,), eps=1e-05, elementwise_affine=True)
      )
    )
    (layernorm_embedding): LayerNorm((16,), eps=1e-05, elementwise_affine=True)
  )
) 
```
and `concrete_args` is:
```
{'head_mask': None, 'decoder_head_mask': None, 'cross_attn_head_mask': None, 'encoder_outputs': None, 'past_key_values': None, 'inputs_embeds': None, 'decoder_inputs_embeds': None, 'use_cache': None, 'output_attentions': None, 'output_hidden_states': None, 'return_dict': None}
```

**Comment by michaelbenayoun** [comment]
I will check on Monday and come back to you, it should be easily fixable I think.

**Comment by donebydan** [comment]
Hey @michaelbenayoun, let me know if you have any thoughts to resolve the tracer issue :)

File: src/transformers/models/bart/modeling_bart.py [hunk 1]
```diff
@@ -128,12 +128,14 @@ def __init__(self, num_embeddings: int, embedding_dim: int):
...
-        """`input_ids_shape` is expected to be [bsz x seqlen]."""
-        bsz, seq_len = input_ids_shape[:2]
+    def forward(self, input_ids: torch.Tensor, past_key_values_length: int = 0):
+        """`input_ids' shape is expected to be [bsz x seqlen]."""
+
+        bsz, seq_len = input_ids.shape[:2]
         positions = torch.arange(
             past_key_values_length, past_key_values_length + seq_len, dtype=torch.long, device=self.weight.device
-        )
+        ).repeat(bsz, 1)
```
Comment (michaelbenayoun) [inline src/transformers/models/bart/modeling_bart.py:137]: After looking into it, it comes from this `repeat`.
The error you get is something like "repeat: list index out of range", because for some reason, during tracing the `repeat` call becomes this:
```
kind = 'call_method'
target = 'repeat'
args = (Proxy(arange), Proxy(getitem_2), 1)
```
It is as if it was considering `self` (here `Proxy(arange)`) as a size in `torch.Tensor.repeat`...
Because this operation is failing, the tracer cannot compute the metadata needed to handle control-flow, hence your error.

In any case, I was able to fix it by doing this:
```suggestion
        ).expand(bsz, -1)
```
Which I think is better anyway because `expand` does not copy data, but maybe you want to copy stuff? I don't have the big picture here.

Comment (donebydan) [reply src/transformers/models/bart/modeling_bart.py]: Thank you so very much for looking into this @michaelbenayoun ! Great find. The use case does not care about copying stuff, so +1 for the more efficient method! :+1:

**Comment by donebydan** [comment]
@sgugger all resolved now. Would you mind giving the PR another look?

File: src/transformers/models/bart/modeling_bart.py [hunk 2]
```diff
@@ -1013,18 +1015,20 @@ def forward(
         if input_ids is not None and inputs_embeds is not None:
             raise ValueError("You cannot specify both decoder_input_ids and decoder_inputs_embeds at the same time")
         elif input_ids is not None:
-            input_shape = input_ids.size()
+            input = input_ids
+            input_shape = input.shape
```
Comment (sgugger) [inline src/transformers/models/bart/modeling_bart.py:1019]: It looks like `input_shape` is not use in this `forward` anymore, and the other variables are not strictly necessary either, so you could should remove those to elif and just do an
`elif input_ids is None and inputs_embeds is  None` for the last `ValueError`.

File: src/transformers/models/bart/modeling_bart.py [hunk 3]
```diff
@@ -1013,18 +1015,20 @@ def forward(
...
+            input = inputs_embeds[:, :, -1]
         else:
             raise ValueError("You have to specify either decoder_input_ids or decoder_inputs_embeds")
 
         # past_key_values_length
         past_key_values_length = past_key_values[0][0].shape[2] if past_key_values is not None else 0
 
         if inputs_embeds is None:
-            inputs_embeds = self.embed_tokens(input_ids) * self.embed_scale
+            inputs_embeds = self.embed_tokens(input) * self.embed_scale
```
Comment (sgugger) [inline src/transformers/models/bart/modeling_bart.py:1031]: Why replace here as `inputs` will be `input_ids`?

File: src/transformers/models/bart/modeling_bart.py [hunk 4]
```diff
@@ -1036,7 +1040,7 @@ def forward(
             encoder_attention_mask = _expand_mask(encoder_attention_mask, inputs_embeds.dtype, tgt_len=input_shape[-1])
 
         # embed positions
-        positions = self.embed_positions(input_shape, past_key_values_length)
+        positions = self.embed_positions(input, past_key_values_length)
```
Comment (sgugger) [inline src/transformers/models/bart/modeling_bart.py:1043]: Looks like it could always be `inputs_embeds[:, :, -1]` here since `input_embeds` are now defined.

**Review by sgugger — COMMENTED** [review]
I think we can simplify a tiny bit more the new code of bart forward, left some suggestions. Apart from that all looks good to me!

File: src/transformers/models/bart/modeling_bart.py (hunk 2 above)
Comment (donebydan) [reply src/transformers/models/bart/modeling_bart.py:1019]: `input_shape` is used for the `_prepare_decoder_attention_mask` and `_expand_mask` methods within this `forward`.

File: src/transformers/models/bart/modeling_bart.py (hunk 3 above)
Comment (donebydan) [reply src/transformers/models/bart/modeling_bart.py:1031]: Just for clarity that `input` is set after the `if, elif` checks. Happy to keep it `input_ids` if you feel strongly about it.

File: src/transformers/models/bart/modeling_bart.py (hunk 4 above)
Comment (donebydan) [reply src/transformers/models/bart/modeling_bart.py:1043]: Perhaps I'm missing something, but if `inputs_embeds` is not provided, the `input` here will be `input_ids`. If `inputs_embeds` _is_ provided (but `input_ids` is not), this `input` will be `inputs_embeds[:, :, -1]`. So the dependency on what is provided means we cannot always pass `inputs_embeds[:, :, -1]`.

File: src/transformers/models/bart/modeling_bart.py (hunk 2 above)
Comment (sgugger) [reply src/transformers/models/bart/modeling_bart.py:1019]: Ah yes, was looking at the wrong forward 🤦‍♂️

File: src/transformers/models/bart/modeling_bart.py (hunk 4 above)
Comment (sgugger) [reply src/transformers/models/bart/modeling_bart.py:1043]: Yes but the `input_embeds` are now defined and their shape will match. Anyhow it's not important, it was to go along with the suggestion above to remove `input` entirely.

**Review by sgugger — APPROVED** [review]

**Comment by sgugger** [comment]
Thanks a lot for working on this!

**Comment by donebydan** [comment]
No problem, thank you for your support 👍
